tests/posix/rw/rw.cpp

```diff
@@ -1011,7 +1064,120 @@ TEST_F(rw, sparse_files)
 	ASSERT_EQ(memcmp(buf + 1, "test", 5), 0);
 	ASSERT_EQ(is_zeroed(buf + 6, 4096 - 6), 1);
 	ASSERT_EQ(memcmp(buf + 4096, "test", 5), 0);
+	EXPECT_TRUE(test_pmemfile_stats_match(pfp, 2, 0, 0, 2));
+
+	/*
+	 * Now that there are no holes, seeking to data should simply just set
+	 * the offset to the given argument, and seeking to hole should seek to
+	 * the end of the file.
+	 */
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 1, PMEMFILE_SEEK_HOLE), size);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 0, PMEMFILE_SEEK_HOLE), size);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4095, PMEMFILE_SEEK_HOLE), size);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4096, PMEMFILE_SEEK_HOLE), size);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4097, PMEMFILE_SEEK_HOLE), size);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, size, PMEMFILE_SEEK_HOLE), size);
+	errno = 0;
+	ASSERT_EQ(pmemfile_lseek(pfp, f, size + 1, PMEMFILE_SEEK_HOLE), -1);
+	ASSERT_EQ(errno, ENXIO);
+
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 1, PMEMFILE_SEEK_DATA), 1);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 0, PMEMFILE_SEEK_DATA), 0);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4094, PMEMFILE_SEEK_DATA), 4094);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4095, PMEMFILE_SEEK_DATA), 4095);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4096, PMEMFILE_SEEK_DATA), 4096);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4097, PMEMFILE_SEEK_DATA), 4097);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 4098, PMEMFILE_SEEK_DATA), 4098);
+	ASSERT_EQ(pmemfile_lseek(pfp, f, size, PMEMFILE_SEEK_DATA), size);
+	errno = 0;
+	ASSERT_EQ(pmemfile_lseek(pfp, f, size + 1, PMEMFILE_SEEK_DATA), -1);
+	ASSERT_EQ(errno, ENXIO);
+
+	/*
+	 * The following tests would become too complicated without assuming
+	 * a fixed block size. Punching holes in files might zero out some
+	 * parts of some blocks, and deallocate some other blocks, while
+	 * seeking to hole/data only seeks to block boundaries.
+	 */
+	if (env_block_size != 0x1000)
+		goto end;
+
+	/*
+	 * Making a hole at the end of the file. The only interface in the API
+	 * that can achieve this is fallocate.
+	 */
+	size = 0x40000;
+	hole = size / 2; /* The hole starts at this offset */
+	r = pmemfile_ftruncate(pfp, f, size);
+	ASSERT_EQ(r, 0) << strerror(errno);
+	r = pmemfile_fallocate(pfp, f,
+			       PMEMFILE_FL_PUNCH_HOLE | PMEMFILE_FL_KEEP_SIZE,
+			       hole, size);
+	ASSERT_EQ(r, 0) << strerror(errno);
```

Comment:
FTR, the code below assumes that ftruncate didn't cheat and allocated all blocks. Maybe it's worth enforcing or at least adding a comment.